sd/endpointer.go

```diff
@@ -0,0 +1,75 @@
+package sd
+
+import (
+	"io"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/log"
+	iEndpoint "github.com/go-kit/kit/sd/internal/endpoint"
+)
+
+// Endpointer listens to a service discovery system and yields a set of
+// identical endpoints on demand. An error indicates a problem with connectivity
+// to the service discovery system, or within the system itself; an Endpointer
+// may yield no endpoints without error.
+type Endpointer interface {
+	Endpoints() ([]endpoint.Endpoint, error)
```

Comment:
I agree that it's important in theory, but this is not how the current implementations behave, e.g. `consul` and `dnssrv`, as examples of internal push and pull models, both return nil for the error from their current `Endpoints()` method. If an implementation uses push model both internally and in the `Instancer` (via `internal/instance.Cache`), how do you see it returning an error from the pull model's `Endpoints()` method? I.e. which error would it be, the last one received via push (such as lost connection to sd backend)?

Note that I had a TBD in the Instancer interface about using richer types in the channels than plain `[]string`. Doing it that way would at least allow the internal sd errors to bubble up to the Instancer subscribers. But it still runs into the same issue as above when push model of the Instancer flips to a pull model of the Endpointer.